checklist/endpoint.go

```diff
@@ -10,96 +10,94 @@ import (
 		req := request.(saveTaskRequest)
 		task := &todo.Task{Name: req.Name}
 		err := s.Save(ctx, task)
-		return saveTaskResponse{task.ID, task.Name, task.Done, err}, nil
+		return saveTaskResponse{taskResponse{task.ID, task.Name, task.Done}, err}, nil
 	}
 
 	return otelkit.EndpointMiddleware(otelkit.WithOperation("saveTask"))(ep)
 }
 
-type listTasksResponse struct {
-	Tasks []struct {
-		ID   int64  `json:"id"`
-		Name string `json:"name"`
-		Done bool   `json:"done"`
-	}
-	Err error `json:"error,omitempty"`
-}
-
-func (r listTasksResponse) MarshalJSON() ([]byte, error) {
-	if r.Failed() == nil {
-		return json.Marshal(r.Tasks)
-	}
-	return json.Marshal(r)
-}
-
-func (r listTasksResponse) Failed() error { return r.Err }
-
 func makeListTasksEndpoint(s Service) endpoint.Endpoint {
 	ep := func(ctx context.Context, _ interface{}) (interface{}, error) {
 		tasks, err := s.List(ctx)
-		var resp = listTasksResponse{Tasks: make([]struct {
-			ID   int64  `json:"id"`
-			Name string `json:"name"`
-			Done bool   `json:"done"`
-		}, 0), Err: err}
+		var resp = listTasksResponse{Tasks: make([]taskResponse, 0), Err: err}
 		if err != nil {
 			return resp, nil
 		}
 		for _, task := range tasks {
-			resp.Tasks = append(resp.Tasks, struct {
-				ID   int64  `json:"id"`
-				Name string `json:"name"`
-				Done bool   `json:"done"`
-			}{ID: task.ID, Name: task.Name, Done: task.Done})
+			resp.Tasks = append(resp.Tasks, taskResponse{ID: task.ID, Name: task.Name, Done: task.Done})
 		}
 		return resp, nil
 	}
 
 	return otelkit.EndpointMiddleware(otelkit.WithOperation("listTasks"))(ep)
 }
 
-type removeTaskRequest struct {
-	id int64
-}
-
 func makeRemoveTaskEndpoint(s Service) endpoint.Endpoint {
 	ep := func(ctx context.Context, request interface{}) (interface{}, error) {
 		req := request.(removeTaskRequest)
-		err := s.Remove(ctx, req.id)
-		return nil, err
+		err := s.Remove(ctx, req.ID)
+		return emptyResponse{Err: err}, nil
 	}
 
 	return otelkit.EndpointMiddleware(otelkit.WithOperation("removeTask"))(ep)
 }
 
-type toggleTaskRequest struct {
-	id int64
-}
-
 func makeToggleTaskEndpoint(s Service) endpoint.Endpoint {
 	ep := func(ctx context.Context, request interface{}) (interface{}, error) {
 		req := request.(toggleTaskRequest)
-		err := s.ToggleDone(ctx, req.id)
-		return nil, err
+		err := s.ToggleDone(ctx, req.ID)
+		return emptyResponse{Err: err}, nil
 	}
 
 	return otelkit.EndpointMiddleware(otelkit.WithOperation("toggleTask"))(ep)
 }
+
+type taskResponse struct {
+	ID   int64  `json:"id"`
+	Name string `json:"name"`
+	Done bool   `json:"done"`
+}
+
+type saveTaskRequest struct {
+	Name string `json:"name"`
+}
+
+type saveTaskResponse struct {
+	taskResponse
+	Err error `json:"error,omitempty"`
+}
+
+func (r saveTaskResponse) Failed() error { return r.Err }
+
+type listTasksResponse struct {
+	Tasks []taskResponse
+	Err   error `json:"error,omitempty"`
+}
+
+func (r listTasksResponse) MarshalJSON() ([]byte, error) {
```

Comment:
that's interesting. defining our own MarshalJSON on our response dto.

Reply:
yep.. this gets used by encoding/json pkg whenever we try to decode this struct